packages/nitro-protocol/contracts/ninja-nitro/readme.md

```diff
@@ -0,0 +1,94 @@
+
+One such quirk is as follows. In Vanilla-Nitro, the hash of the status of a channel (which includes the full outcome, the address of any challenger, the time when the channel will finalize, etc) is stored in a mapping, keyed by the `channelId`.
+
+In Ninja-Nitro, exactly the same is true: only each proxy has its own mapping. The reason that this is quirky is that each of these mappings is only ever going to have a single entry, because each proxy is in a 1-1 relation to a `channelId`. The reason that this quirk is not actually very troublesome are:
+
+- the gas cost are not dramatically higher than storing a single hash without using a mapping (investigations suggest around 100 gas difference). See https://docs.soliditylang.org/en/v0.8.3/types.html?highlight=mapping#mapping-types
+- we are able to reuse `ForceMove` (in the sense of `SingleChannelAdjudicator is ForceMove`) in its entirety without modifying it at all
+
+> A future optimization might be to replace the mapping with a regular, singleton storage variable. This would require a small but breaking change to `ForceMove.sol`.
+
+### External destinations
+
+In Vanilla-Nitro, funds for channels are tracked _internally_, while funds for any Ethereum address (e.g. an externally owned account) are typically paid out of the system when defunding a channel. Padding an Ethereum address with enough zeros results in a 32 byte number which we call an external destination. Channel ids are known as internal destinations.
+
+In Ninja-Nitro, funds are always tracked externally. This is possible since (up to hash collisions) each channel now has a unique Ethereum address. Funds are therefore always paid _out_ of a `Proxy` contract after a `transfer` or `claim`. There is therefore no such thing as an internal destination in Ninja-Nitro. Everything is external.
+
+To maintain compatibility with the off-chain code we already have for Vanilla-Nitro, the Ninja-Nitro implementation converts destinations to addresses in the following way:
+
+- if the destination is a padded Ethereum address, slice off the Ethereum address and use that
+- if the destination is not a padded Ethereum address, compute the create2 address of that channel and use that
+
+> A future optimization might be to do away with the 32 byte destination concept altogether. Destinations would then always be Ethereum addresses, and trying to allocate funds to a 32 byte destination would raise an exception.
+
+## Gas tradeoffs
+
+| Operation                                                      | Vanilla | Ninja |
+| -------------------------------------------------------------- | ------- | ----- |
+| infastructure deployment                                       | 4.3M    | 5.1M  |
+| directly-funded unidirectional payment channel (happy path)    | 212K    | 167K  |
```

Comment:
This row needs quite a bit of context to make sense of this:
- how many assets? what asset types?
- if this channel has an ERC20 asset, does Vanilla include the `approve` transaction in the 212k? Should it?
- for a fair comparison, should we take into account the gas savings in Vanilla from doing https://github.com/statechannels/statechannels/pull/3394?